src/hotspot/share/oops/method.cpp

```diff
@@ -1290,12 +1292,21 @@ address Method::verified_code_entry() {
 // Not inline to avoid circular ref.
 bool Method::check_code() const {
   // cached in a register or local.  There's a race on the value of the field.
-  CompiledMethod *code = Atomic::load_acquire(&_code);
-  return code == NULL || (code->method() == NULL) || (code->method() == (Method*)this && !code->is_osr_method());
+  CodeBlob *blob = Atomic::load_acquire(&_code);
+  if (blob == nullptr) {
+    return true;
+  } else if (blob->is_compiled()) {
+    CompiledMethod* cm = blob->as_compiled_method();
+    return (cm->method() == nullptr) || (cm->method() == this && !cm->is_osr_method());
+  } else {
+    assert(blob->is_mh_intrinsic(), "must be MH intrinsic");
+    MethodHandleIntrinsicBlob* mhi = blob->as_mh_intrinsic();
+    return (mhi->method() == nullptr) || (mhi->method() == this);
```

Comment:
The assert looks redundant, since the cast on the next line already checks it.

Also, can `mhi->method()` really be `nullptr` here?